### PR TITLE
[TrimmableTypeMap] Add TypeMap proxy groundwork with CreateInstance

### PR DESCRIPTION
**Pull request by simonrozsival**
Part of https://github.com/dotnet/android/issues/10799

## Summary

Adds the core pipeline that transforms scanned `JavaPeerInfo` records into TypeMap `.dll` assemblies with proxy types and `CreateInstance` support.

### Generator pipeline

```
   Input assemblies (*.dll)
           │
           ▼
   ┌───────────────┐
   │ JavaPeerScanner│  Scans PE metadata for [Register] types,
   └───────┬───────┘  activation ctors, interfaces, base types
           │
           │  List
           ▼
   ┌───────────────┐
   │  ModelBuilder  │  Alias detection, ACW filtering, proxy naming,
   └───────┬───────┘  unconditional vs trimmable classification
           │
           │  TypeMapAssemblyData
           ▼
┌──────────────────────┐
│ TypeMapAssemblyEmitter│  IL emission: proxy types, CreateInstance,
│   + PEAssemblyBuilder │  TypeMap attributes, IgnoresAccessChecksTo
└──────────┬───────────┘
           │
           │  PE bytes
           ▼
   ┌───────────────┐
   │  Output *.dll  │  _MyApp.TypeMap.dll (per-assembly)
   └───────────────┘  _Microsoft.Android.TypeMaps.dll (root)
```

### Tests
- 141 tests covering model builder, assembly generator, root assembly generator, and scanner behavior
- Test fixtures include MCW bindings, user ACW types, interfaces with invokers, nested types, generics, Implementor/EventDispatcher helpers, JI-style activation ctors, and edge cases

### Future optimization
- Implementor and EventDispatcher types are currently treated as unconditional (matching legacy behavior). Trimming optimization deferred to #10911.

UCO wrappers, RegisterNatives, and IAndroidCallableWrapper support will be added in follow-up PRs.

## Follow-up work

- **#10913** — Integration tests for generator-consumed scanner fields (`IsUnconditional`, `InvokerTypeName`, `CompatJniName`, `ManagedMethodName`). These are new concepts with no legacy equivalent, so they cannot be validated via legacy-vs-new comparison and need standalone correctness tests.
- **#10807** — Decompose `_GenerateJavaStubs`: shared tasks for trimmable path
- **#10911** — Trimming optimization for Implementor/EventDispatcher types

### CODE REVIEW

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs [hunk 1]
```diff
@@ -732,4 +696,28 @@ static string GetCrc64PackageName (string ns, string assemblyName)
...
+	{
+		int lastDot = fullName.LastIndexOf ('.');
+		string typePart = lastDot >= 0 ? fullName.Substring (lastDot + 1) : fullName;
+		int lastPlus = typePart.LastIndexOf ('+');
+		return lastPlus >= 0 ? typePart.Substring (lastPlus + 1) : typePart;
+	}
+
+	static List<JniParameterInfo> ParseJniParameters (string jniSignature)
+	{
+		var typeStrings = JniSignatureHelper.ParseParameterTypeStrings (jniSignature);
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs]: I wonder if `JniSignatureHelper.ParseParameterTypeStrings` should just return `List<JniParameterInfo>`? We could avoid looping over the collection one extra time. There doesn't seem to be any other call to `ParseParameterTypeStrings`, right?

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs [hunk 2]
```diff
@@ -732,4 +696,28 @@ static string GetCrc64PackageName (string ns, string assemblyName)
 		var hash = System.IO.Hashing.Crc64.Hash (data);
 		return $"crc64{BitConverter.ToString (hash).Replace ("-", "").ToLowerInvariant ()}";
 	}
+
+	static string ExtractNamespace (string fullName)
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs:738]: Could we just return `ReadOnlySpan<char>`? Would that be beneficial in the locations where we call this method?

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs [hunk 3]
```diff
@@ -732,4 +696,28 @@ static string GetCrc64PackageName (string ns, string assemblyName)
...
+	{
+		int lastDot = fullName.LastIndexOf ('.');
+		return lastDot >= 0 ? fullName.Substring (0, lastDot) : "";
+	}
+
+	static string ExtractShortName (string fullName)
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs:747]: Could we just return `ReadOnlySpan<char>`? Would that be beneficial in the locations where we call this method?

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs [hunk 4]
```diff
@@ -274,62 +268,32 @@ static void AddMarshalMethod (List<MarshalMethodInfo> methods, RegisterInfo regi
...
+		bool isConstructor = registerInfo.JniName == "<init>" || registerInfo.JniName == ".ctor";
+		string nativeCallbackName = $"n_{index.Reader.GetString (methodDef.Name)}";
+		if (isConstructor) {
+			int ctorIndex = 0;
+			foreach (var method in methods) {
+				if (method.IsConstructor) {
+					ctorIndex++;
+				}
+			}
+			nativeCallbackName = ctorIndex == 0 ? "n_ctor" : $"n_ctor_{ctorIndex}";
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerScanner.cs]: Do we need the counter implemented this way? Could we instead maybe just use the method hash code as a unique identifier?

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+			},
+		};
+		return peer;
+	}
+
+	protected static JavaPeerInfo MakeInterfacePeer (
+		string jniName = "android/view/View$OnClickListener",
+		string managedName = "Android.Views.View+IOnClickListener",
+		string asmName = "Mono.Android",
+		string invokerName = "Android.Views.View+IOnClickListenerInvoker")
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs]: I don't think this method should have default parameter values. If we want to have examples, let's use doc comments.

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+			ManagedTypeName = managedName,
+			ManagedTypeNamespace = ns,
+			ManagedTypeShortName = shortName,
+			AssemblyName = asmName,
+			IsInterface = true,
+			InvokerTypeName = invokerName,
+		};
+	}
+
+	protected static MarshalMethodInfo MakeMarshalMethod (string jniName, string callbackName, string jniSig, bool isConstructor = false)
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs]: This method doesn't seem useful. It just forwards the parameters to the properties. Let's just inline the `new` call instead of using this method.

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+///
+/// // Emitted so the proxy assembly can access internal members in the target assembly:
+/// [assembly: IgnoresAccessChecksTo("Mono.Android")]
+/// </code>
+/// </remarks>
+sealed class TypeMapAssemblyEmitter
+{
+	readonly Version _systemRuntimeVersion;
+
+	PEAssemblyBuilder _pe = null!;
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs]: I'm not a fan of `null!`. Let's make this `readonly` and initialize `_pe` in the constructor.

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+	}
+
+	void EmitCreateInstance (JavaPeerProxyData proxy)
+	{
+		if (!proxy.HasActivation) {
+			EmitCreateInstanceBody (encoder => {
+				encoder.OpCode (ILOpCode.Ldnull);
+				encoder.OpCode (ILOpCode.Ret);
+			});
+			return;
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs]: Extract this into a separate method

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+
+		// Generic type definitions cannot be instantiated
+		if (proxy.IsGenericDefinition) {
+			EmitCreateInstanceBody (encoder => {
+				encoder.LoadString (_pe.Metadata.GetOrAddUserString ("Cannot create instance of open generic type."));
+				encoder.OpCode (ILOpCode.Newobj);
+				encoder.Token (_notSupportedExceptionCtorRef);
+				encoder.OpCode (ILOpCode.Throw);
+			});
+			return;
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs]: Extract this into a separate method

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+		// Interface with invoker: new TInvoker(IntPtr, JniHandleOwnership)
+		if (proxy.InvokerType != null) {
+			var invokerCtorRef = AddActivationCtorRef (_pe.ResolveTypeRef (proxy.InvokerType));
+			EmitCreateInstanceBody (encoder => {
+				encoder.OpCode (ILOpCode.Ldarg_1);
+				encoder.OpCode (ILOpCode.Ldarg_2);
+				encoder.OpCode (ILOpCode.Newobj);
+				encoder.Token (invokerCtorRef);
+				encoder.OpCode (ILOpCode.Ret);
+			});
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs]: Extract this into a separate method

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+		if (activationCtor.IsOnLeafType) {
+			// Leaf type has its own ctor: new T(IntPtr, JniHandleOwnership)
+			var ctorRef = AddActivationCtorRef (targetTypeRef);
+			EmitCreateInstanceBody (encoder => {
+				encoder.OpCode (ILOpCode.Ldarg_1);
+				encoder.OpCode (ILOpCode.Ldarg_2);
+				encoder.OpCode (ILOpCode.Newobj);
+				encoder.Token (ctorRef);
+				encoder.OpCode (ILOpCode.Ret);
+			});
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs]: Extract this into a separate method

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+				encoder.OpCode (ILOpCode.Castclass);
+				encoder.Token (targetTypeRef);
+
+				encoder.OpCode (ILOpCode.Dup);
+				encoder.OpCode (ILOpCode.Ldarg_1);
+				encoder.OpCode (ILOpCode.Ldarg_2);
+				encoder.Call (baseActivationCtorRef);
+
+				encoder.OpCode (ILOpCode.Ret);
+			});
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs]: Extract this into a separate method

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/RootTypeMapAssemblyGeneratorTests.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+	public void Dispose () => DeleteTempDir (_outputDir);
+
+	string GenerateRootAssembly (IReadOnlyList<string> perAssemblyNames, string? assemblyName = null)
+	{
+		var outputPath = Path.Combine (_outputDir,
+			(assemblyName ?? "_Microsoft.Android.TypeMaps") + ".dll");
+		var generator = new RootTypeMapAssemblyGenerator (new Version (11, 0, 0, 0));
+		generator.Generate (perAssemblyNames, outputPath, assemblyName);
+		return outputPath;
+	}
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/RootTypeMapAssemblyGeneratorTests.cs]: I'm not happy that this test writes a file to disk. Can we just write the data to a `Stream` and use the `PEReader` methods to read from `Stream` instead of from a file path. This should make tests more reliable and avoid weird/flaky IO errors eventually.

**Review by simonrozsival — COMMENTED** [review]
@copilot address these comments

File: build-tools/automation/azure-pipelines-public.yaml [hunk 14]
```diff
@@ -162,6 +162,7 @@ stages:
         buildResultArtifactName: Build Results - Linux
         xaSourcePath: $(System.DefaultWorkingDirectory)/android
         nugetArtifactName: $(LinuxNuGetArtifactName)
+        makeMSBuildArgs: -m:2
```
Comment (jonathanpeppers) [inline build-tools/automation/azure-pipelines-public.yaml]: Is this leftover from something? Can we remove it?

What is odd, this is Linux and it won't have file locking issues like Windows would.

**Review by jonathanpeppers — COMMENTED** [review]
My AI code reviewer says this, but I don't think we need to take any action except look at the yaml changes:

```
 PR #10808 Review Summary

  Title: [TrimmableTypeMap] Add TypeMap proxy groundwork with CreateInstance
  Author: @simonrozsival | +2956 / -88 across 20 files | 43 commits

  What it does

  Adds the core code generation pipeline that transforms scanned JavaPeerInfo records into per-assembly TypeMap .dll assemblies containing proxy types with CreateInstance methods. Pipeline: Scanner →
  ModelBuilder → Emitter → PE output.

  Architecture (excellent)

  Clean 3-layer separation:

   - ModelBuilder — all decision logic (dedup, alias detection, ACW filtering, 2-arg vs 3-arg classification). Pure data in, pure data out.
   - TypeMapAssemblyEmitter — mechanical IL emission from the model. No business logic.
   - PEAssemblyBuilder — shared plumbing (MetadataBuilder, assembly refs, scratch BlobBuilders, PE serialisation). Reused by both per-assembly and root generators.

  Findings

  ┌───┬────────────┬─────────────────────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ # │ Severity   │ File                                │ Issue                                                                                                                                               │
  ├───┼────────────┼─────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ 1 │ Medium     │ JavaPeerInfo.cs                     │ ManagedTypeNamespace and ManagedTypeShortName are added and populated by the scanner, but not consumed by any production code in this PR. The       │
  │   │            │                                     │ ModelBuilder and emitter don't reference them. Likely for follow-up PRs (UCO wrappers / RegisterNatives), but this is speculative code shipping     │
  │   │            │                                     │ ahead of its consumer.                                                                                                                              │
  ├───┼────────────┼─────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ 2 │ Low        │ TypeMapAssemblyData.cs              │ Contains 6 types in one file (170 lines). The types are cohesive model records, but one-type-per-file is a strict convention in this org. Could     │
  │   │            │                                     │ split into TypeMapAttributeData.cs, JavaPeerProxyData.cs, TypeRefData.cs, ActivationCtorData.cs, TypeMapAssociationData.cs.                         │
  ├───┼────────────┼─────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ 3 │ Low        │ TypeMapAssemblyGeneratorTests.cs,   │ Missing newline at end of file (\ No newline at end of file).                                                                                       │
  │   │            │ TypeMapModelBuilderTests.cs         │                                                                                                                                                     │
  └───┴────────────┴─────────────────────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

  Positives

   - Deterministic output: SortedDictionary for JNI name ordering, sorted aliases, content-fingerprinted MVID — builds are reproducible.
   - PEAssemblyBuilder scratch BlobBuilders: Reuses _sigBlob/_codeBlob/_attrBlob per method, avoiding per-emission allocations. Correctly captures the sig blob handle before emitIL callbacks (which can call
  AddMemberRef and clear _sigBlob). There's even a regression test (EmitBody_ILCallbackCallsAddMemberRef_SignatureNotCorrupted) for this.
   - 141 tests covering model builder, PE generation round-trips, root assembly, scanner behavior, and edge cases (generics, nested types, aliases, JI-style ctors, invoker exclusion).
   - Test infrastructure refactoring: Extracted FixtureTestBase with Lazy<> cached fixture scanning — scanner tests and generator tests share the same fixture without re-scanning.
   - IL emission comments map to pseudo-C# equivalents, making the ECMA-335 byte-level code readable.
   - 5 CreateInstance strategies properly handled: leaf newobj, inherited (GetUninitializedObject), invoker, JI-style (with JniObjectReference conversion + DeleteRef cleanup), generic definition (throws), and
   no-activation (returns null).

  Summary

  This is a well-structured, well-tested foundation PR. The main substantive concern is the two unused JavaPeerInfo properties shipping without consumers. The rest is clean.
```
